[inductor] Allow sympy expressions to participate in type promotion

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115676
* #115700
* #115699
* #115677

In the test example we have `add(i64[10], sympy.Expr)` where
`sympy.Expr` is not considered a promoting arg so isn't factored into
the type promotion. However, in eager it would promote to float32.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 committed:
[inductor] Allow sympy expressions to participate in type promotion

In the test example we have `add(i64[10], sympy.Expr)` where
`sympy.Expr` is not considered a promoting arg so isn't factored into
the type promotion. However, in eager it would promote to float32.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Allow sympy expressions to participate in type promotion"


In the test example we have `add(i64[10], sympy.Expr)` where
`sympy.Expr` is not considered a promoting arg so isn't factored into
the type promotion. However, in eager it would promote to float32.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -4588,6 +4609,14 @@ def fn(x):
+    def test_float_index_expression_type_promotion(self):
+        # Test that float indexing expressions participate in type promotion
+        def fn(x):
+            return x + 1.0 / x.size(0)
```
If `x` is not defined inside `fn`, `x` will not be seen as an indexing expression

peterbell10 replied:
`x.size(0)` is the indexing expression here, `x` is just a tensor.

peterbell10 committed:
Update on "[inductor] Allow sympy expressions to participate in type promotion"


In the test example we have `add(i64[10], sympy.Expr)` where
`sympy.Expr` is not considered a promoting arg so isn't factored into
the type promotion. However, in eager it would promote to float32.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2708,7 +2708,8 @@ def forward(self, idx, x):
         with torch.no_grad():
             metrics.reset()
-            self.common(m, (idx, x))
+            # FIXME: returns the wrong dtype
```
Can you open a tracking issue for these, or will you submit fixes for them?

lezcano replied:
nvm, you already did

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -198,7 +198,9 @@ def transform_args(args, broadcast, type_promotion_kind, convert_input_to_bool):
         else:
             # FIXME that's a crude approximation for promoting args
```
At some point we should look into this comment lol

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge
